### DISCUSSION TITLE
Regarding elastic_strain and plastic_strain

### DISCUSSION THREAD

**Discussion by RavisCPP** [idaholab/moose#26573] [Q&A Modules: Solid mechanics]
Hi everyone, I am trying to use "FiniteStrainUOBasedCP" for one of my problems, and I am able to get outputs for all the stress and strain components. But as soon as I try to get the plastic_strain_zz, effective_plastic_strain and elastic_strain_zz for my analysis, it throws an error:

"The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'plastic_strain', requested by 'plastic_strain_zz_all' is not defined on block 0
Material property 'plastic_strain', requested by 'plastic_strain_zz_all_face' is not defined on block 0
Material property 'plastic_strain', requested by 'plastic_strain_zz_all_neighbor' is not defined on block 0"

I am also attaching my input files, in this, as per TensorMechanicsMaster syntax, I added elastic_strain_zz and plastic_strain_zz in the generate_output option. If I do not add these two, other outputs are getting generated normally!

[input_slip_sys.txt](https://github.com/idaholab/moose/files/13962475/input_slip_sys.txt)
 
[euler_ang_file.txt](https://github.com/idaholab/moose/files/13962476/euler_ang_file.txt)

[save_euler_1.i.txt](https://github.com/idaholab/moose/files/13962464/save_euler_1.i.txt)

**Comment by GiudGiud** [comment]
Hello

Can you use the Debug/show material prop parameter to check the names of the material properties?

Once you have done that, it s likely you cannot access that material property with the suffix for the components. 

You need to use some sort of auxkernel that will use the regular name of the material (RankTwoAux is an example) and ask you for the components you want to output

Guillaume

**Comment by RavisCPP** [comment]
Hi, I tried the debugging "show_material_props = true" and it gave me a list of all the available material properties. It had the plastic_strain_zz too. Again, same error 
Material property 'plastic_strain', requested by 'plastic_strain' is not defined on block 0
Material property 'plastic_strain', requested by 'plastic_strain_zz_all' is not defined on block 0

So, as per my understanding, just to ckech, I also tried to take output of elastic strain as:
generate_output = 'stress_xx stress_yy stress_zz stress_xy stress_yz stress_xz strain_xx strain_yy strain_zz strain_xy strain_yz strain_xz elastic_strain_zz'

So, it didn't throw any error but the value of elastic strain was 0 everywhere. I think maybe this is happening because plastic_strain is not defined or developed for "FiniteStrainUObasedCP" and same for elastic_strain, but since it was defined at some base class, so, it doesn't throw the error.

[save_euler_2.i.txt](https://github.com/idaholab/moose/files/13976311/save_euler_2.i.txt)

I have also tried implementing elastic_strain and plastic_strain in UOBasedCP, but I have doubts if that is correct or not:

elastic_strain:

void
FiniteStrainUOBasedCP::calcResidual()
{

  //RankTwoTensor iden(RankTwoTensor::initIdentity), ce, ee, ce_pk2, eqv_slip_incr, pk2_new;
  RankTwoTensor iden(RankTwoTensor::initIdentity), ce, ce_pk2, eqv_slip_incr, pk2_new;

  getSlipRates();
  if (_err_tol)
    return;
  
  for (unsigned int i = 0; i < _num_uo_slip_rates; ++i)
    {
      for (unsigned int j = 0; j < _uo_slip_rates[i]->variableSize(); ++j)
      {
        eqv_slip_incr += (*_flow_direction[i])[_qp][j] * (*_mat_prop_slip_rates[i])[_qp][j] * _dt;
      }
    }

  eqv_slip_incr = iden - eqv_slip_incr;
  _fp_inv = _fp_old_inv * eqv_slip_incr;
  _fe = _dfgrd_tmp * _fp_inv;

  ce = _fe.transpose() * _fe;
  _elastic_str[_qp] = ce - iden;
  _elastic_str[_qp]*= 0.5;

  //pk2_new = _elasticity_tensor[_qp] * ee;
  
  pk2_new = _elasticity_tensor[_qp] * _elastic_str[_qp];

  _resid = _pk2[_qp] - pk2_new;
 
}

So, it had elastic_strain defined as ee in local variable, I defined a global variable and defined the elastic_strain like that.

For plastic_strain, I did this:

_pl_str = _pl_str_old;
for (unsigned int i = 0; i < _num_uo_slip_rates; ++i)
    {
      for (unsigned int j = 0; j < _uo_slip_rates[i]->variableSize(); ++j)
      {
        plastic_vel_grad = plastic_vel_grad + (*_flow_direction[i])[_qp][j] * (*_mat_prop_slip_rates[i])[_qp][j] * _dt;
      }
    }
plastic_stretching = 0.5 * (plastic_vel_grad + plastic_vel_grad.transpose());

for (unsigned int i = 0; i < _num_uo_slip_rates; ++i)
    {
      for (unsigned int j = 0; j < _uo_slip_rates[i]->variableSize(); ++j)
      {
        flow_dir = (*_flow_direction[i])[_qp][j];
        flow_dir_t = flow_dir.transpose();
        flow_dir_symm = 0.5 * (flow_dir + flow_dir_t);
        plastic_strain_crystal_increment = plastic_strain_crystal_increment + ((*_mat_prop_slip_rates[i])[_qp][j] * flow_dir_symm) * _dt;
      }
    }
_pl_str = _pl_str + plastic_strain_crystal_increment;

As per my understanding. Is the implementation correct? Or if there is any other way to get plastic strain, it will be really helpful. I am new to moose, but hope to work more towards its development.

**Reply by GiudGiud** [reply]
@recuero
